crates/bevy_ecs/src/system/function_system.rs

```diff
@@ -355,6 +355,22 @@ where
     marker: PhantomData<fn() -> Marker>,
 }
 
+impl<Marker, F> FunctionSystem<Marker, F>
+where
+    F: SystemParamFunction<Marker>,
+{
+    /// Retrieves this systems [`SystemParam::State`] mutably
+    ///
+    /// # Safety
+    /// Mutable reference must not be used to put the [`SystemParam::State`]
+    /// in an invalid state for this [`FunctionSystem`]
```

Comment:
I would like more clarity here if possible. Ideally we can point to exactly where the invariants are documented, and which ones are memory safety concerns.

Reply:
The tricky part here is this function allows you to modify any system param state, including user defined system params. That means that we can't really know all the invariants that need to be maintained here. We could try and list invariants for existing system params though. 

Exposing the state directly like this is a little hacky. My preferred solution would be a macro that generates syntax like this:
```rust
system.set_state::<0>(query_state)
```
But that doesn't solve the problem that allowing users to access system state can be quite unsafe. I'm open to a safer approach but I'm not sure what that looks like.